MathLib/LinAlg/SimpleMatrixVectorProvider-impl.h

```diff
@@ -0,0 +1,281 @@
+            " This might not be a serious error for the current implementation,"
+            " but it might become one in the future."
+            " Hence, I will abort now.");
+        std::abort();
+    }
+
+    if (do_search)
+    {
+        auto it = unused_map.find(id);
+        if (it != unused_map.end()) // unused matrix/vector found
+            return { ::detail::transfer(unused_map, used_map, it), false };
+    }
+
+    // not searched or not found, so create a new one
+    id = _next_id++;
+    auto res = used_map.emplace(
+        MatrixVectorTraits<MatVec>::newInstance(std::forward<Args>(args)...).release(),
+        id);
+    assert(res.second && "Emplacement failed.");
+    return { res.first->first, true };
+}
+
+template<typename Matrix, typename Vector>
+template<bool do_search, typename... Args>
+std::pair<Matrix*, bool>
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix_(std::size_t& id, Args&&... args)
+{
+    return get_<do_search>(id, _unused_matrices, _used_matrices, std::forward<Args>(args)...);
+}
+
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix()
+{
+    std::size_t id = 0u;
+    return *getMatrix_<false>(id).first;
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(std::size_t& id)
+{
+    return *getMatrix_<true>(id).first;
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(MatrixSpecificationsProvider const& msp)
+{
+    std::size_t id = 0u;
+    auto const mat_spec = msp.getMatrixSpecifications();
+    return *getMatrix_<false>(id, mat_spec).first;
+    // TODO assert that the returned object always is of the right size
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(MatrixSpecificationsProvider const& msp, std::size_t& id)
+{
+    auto mat_spec = msp.getMatrixSpecifications();
+    return *getMatrix_<true>(id, mat_spec).first;
+    // TODO assert that the returned object always is of the right size
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(Matrix const& A)
+{
+    std::size_t id = 0u;
+    auto const& res = getMatrix_<false>(id, A);
+    if (!res.second) // no new object has been created
+        BLAS::copy(A, *res.first);
+    return *res.first;
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(Matrix const& A, std::size_t& id)
+{
+    auto const& res = getMatrix_<false>(id, A);
+    if (!res.second) // no new object has been created
+        BLAS::copy(A, *res.first);
+    return *res.first;
+}
+
+template<typename Matrix, typename Vector>
+void
+SimpleMatrixVectorProvider<Matrix, Vector>::
+releaseMatrix(Matrix const& A)
+{
+    auto it = _used_matrices.find(const_cast<Matrix*>(&A));
```

Comment:
Now you almost got me. ^^
Taking the address of a const ref returns a pointer to a const, but `_used_matrices` stores pointers to non-const, so `find` expects the latter. The automatic conversion between them fails.